### ISSUE TITLE
[0.10.0] Nodes joins as meta node instead of data nodes in cluster of more than 4 nodes

### ISSUE THREAD

**Issue by supershal** [influxdata/influxdb#5601]
I am trying to run influxdb cluster of more than 4 nodes. The resulting cluster adds newer node as meta node instead of data node. The cluster only have one data node (first node) and 4 meta node instead of atleast 3 meta nodes.
All nodes in the cluster is getting ID value as 1 as per meta/node.json
Influxdb version: 0.10.0-1
OS: 4.2.2-coreos-r2
My config file is the default config file generated from "influxd config" command.
Binding Address are overridden using environment variables as follows. 
The cluster is created using docker containers on Kubernetes.

Following are steps and results:
`
-----------------------------------------------------------------------------
Node 1:
-----------------------------------------------------------------------------
++ hostname -i
+ HOSTIP=10.244.1.4
+ export INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.1.4:8091
+ INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.1.4:8091
+ export INFLUXDB_META_BIND_ADDRESS=10.244.1.4:8088
+ INFLUXDB_META_BIND_ADDRESS=10.244.1.4:8088
+ influxd --config /etc/influxdb.toml

Node 1 - node.json
-----------------------------------------
root@influxdb-raft-rc-bg2hr:/data/meta# cat node.json
{"ID":1,"MetaServers":["10.244.1.4:8091","10.244.2.188:8091","10.244.2.205:8091","10.244.1.190:8091"]}

Node 1 - CLI:
-----------------------------------------
```
> show servers
name: data_nodes
----------------
id	http_addr	tcp_addr
1	localhost:8086	10.244.1.4:8088

name: meta_nodes
----------------
id	http_addr	tcp_addr
1	10.244.1.4:8091	10.244.1.4:8088
```

-----------------------------------------------------------------------------
Node 2:
-----------------------------------------------------------------------------
++ INFLUXD_OPTS='-join 10.244.1.4:8091'
++ hostname -i
+ HOSTIP=10.244.2.188
+ export INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.2.188:8091
+ INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.2.188:8091
+ export INFLUXDB_META_BIND_ADDRESS=10.244.2.188:8088
+ INFLUXDB_META_BIND_ADDRESS=10.244.2.188:8088
+ influxd --config /etc/influxdb.toml -join 10.244.1.4:8091

Node 2 - node.json
-----------------------------------------
root@influxdb-raft-rc-pdfpe:/data/meta# cat node.json
{"ID":1,"MetaServers":["10.244.1.4:8091","10.244.2.188:8091","10.244.2.205:8091","10.244.1.190:8091"]}

Node 2 - CLI: 
-----------------------------------------
```
> show servers
name: data_nodes
----------------
id	http_addr	tcp_addr
1	localhost:8086	10.244.1.4:8088


name: meta_nodes
----------------
id	http_addr		tcp_addr
1	10.244.1.4:8091		10.244.1.4:8088
2	10.244.2.188:8091	10.244.2.188:8088

```
-----------------------------------------------------------------------------
Node 3:
-----------------------------------------------------------------------------
++ INFLUXD_OPTS='-join 10.244.1.4:8091,10.244.2.188:8091'
++ hostname -i
+ HOSTIP=10.244.2.205
+ export INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.2.205:8091
+ INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.2.205:8091
+ export INFLUXDB_META_BIND_ADDRESS=10.244.2.205:8088
+ INFLUXDB_META_BIND_ADDRESS=10.244.2.205:8088
+ influxd --config /etc/influxdb.toml -join 10.244.1.4:8091,10.244.2.188:8091

Node 3 - Node.json:
-----------------------------------------
root@influxdb-raft-rc-5zyef:/data/meta# cat node.json
{"ID":1,"MetaServers":["10.244.1.4:8091","10.244.2.188:8091","10.244.2.205:8091","10.244.1.190:8091"]}

Node 3 - CLI:
-----------------------------------------
```
> show servers
name: data_nodes
----------------
id	http_addr	tcp_addr
1	localhost:8086	10.244.1.4:8088

name: meta_nodes
----------------
id	http_addr		tcp_addr
1	10.244.1.4:8091		10.244.1.4:8088
2	10.244.2.188:8091	10.244.2.188:8088
3	10.244.2.205:8091	10.244.2.205:8088
```
-----------------------------------------------------------------------------
Node 4:
-----------------------------------------------------------------------------

++ INFLUXD_OPTS='-join 10.244.2.205:8091,10.244.1.4:8091,10.244.2.188:8091'
++ hostname -i
+ HOSTIP=10.244.1.190
+ export INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.1.190:8091
+ INFLUXDB_META_HTTP_BIND_ADDRESS=10.244.1.190:8091
+ export INFLUXDB_META_BIND_ADDRESS=10.244.1.190:8088
+ INFLUXDB_META_BIND_ADDRESS=10.244.1.190:8088
+ influxd --config /etc/influxdb.toml -join 10.244.2.205:8091,10.244.1.4:8091,10.244.2.188:8091

Node 4: node.json
-----------------------------------------
root@influxdb-peers-rc-75jd8:/data/meta# cat node.json
{"ID":1,"MetaServers":["10.244.1.4:8091","10.244.2.188:8091","10.244.2.205:8091","10.244.1.190:8091"]}

Node 4 - CLI:
-----------------------------------------
```
> show servers
name: data_nodes
----------------
id	http_addr	tcp_addr
1	localhost:8086	10.244.1.4:8088

name: meta_nodes
----------------
id	http_addr		tcp_addr
1	10.244.1.4:8091		10.244.1.4:8088
2	10.244.2.188:8091	10.244.2.188:8088
3	10.244.2.205:8091	10.244.2.205:8088
4	10.244.1.190:8091	10.244.1.190:8088

```
-----------------------------------------------------------------------------
Result of cluster query shows only first host(influxdb-raft-rc-bg2hr)
-----------------------------------------------------------------------------

> select * from cluster where time > now() - 3h
name: cluster
-------------
time				clusterID			hostname			mapShardReq	mapShardResp	nodeID
1455045073000000000	3863944593121881570	influxdb-raft-rc-bg2hr	1		1		0
1455045083000000000	3863944593121881570	influxdb-raft-rc-bg2hr	2		3		0
1455045093000000000	3863944593121881570	influxdb-raft-rc-bg2hr	2		3		0
1455045103000000000	3863944593121881570	influxdb-raft-rc-bg2hr	2		3		0
1455045113000000000	3863944593121881570	influxdb-raft-rc-bg2hr	2		3		0
1455045123000000000	3863944593121881570	influxdb-raft-rc-bg2hr	3		5		0
1455045133000000000	3863944593121881570	influxdb-raft-rc-bg2hr	4		6		0
1455045143000000000	3863944593121881570	influxdb-raft-rc-bg2hr	4		6		0
1455045153000000000	3863944593121881570	influxdb-raft-rc-bg2hr	5		8		0
1455045163000000000	3863944593121881570	influxdb-raft-rc-bg2hr	5		8		0
`

**Comment by rossmcdonald** [comment]
@supershal Please try the solution from #5585, as I think this is the same issue.

**Comment by supershal** [comment]
Thanks @rossmcdonald . I tried suggested workaround. Exported following variables before starting influxd
```
HOSTIP=$(hostname -i)
export INFLUXDB_META_HTTP_BIND_ADDRESS=$HOSTIP:8091
export INFLUXDB_META_BIND_ADDRESS=$HOSTIP:8088
export INFLUXDB_HTTP_BIND_ADDRESS=$HOSTIP:8086
```
Now I see all instances in both meta nodes and data nodes with different ids for the same node.
```
> show servers
name: data_nodes
----------------
id	http_addr		tcp_addr
1	10.244.1.7:8086		10.244.1.7:8088
4	10.244.2.190:8086	10.244.2.190:8088
5	10.244.1.179:8086	10.244.1.179:8088
7	10.244.2.194:8086	10.244.2.194:8088
9	10.244.3.237:8086	10.244.3.237:8088
11	10.244.2.66:8086	10.244.2.66:8088


name: meta_nodes
----------------
id	http_addr		tcp_addr
1	10.244.1.7:8091		10.244.1.7:8088
2	10.244.2.190:8091	10.244.2.190:8088
3	10.244.1.179:8091	10.244.1.179:8088
6	10.244.2.194:8091	10.244.2.194:8088
8	10.244.3.237:8091	10.244.3.237:8088
10	10.244.2.66:8091	10.244.2.66:8088
```

However I do see all members reporting in "cluster" measurement table.
```
> show tag values from cluster with key=hostname
name: hostnameTagValues
-----------------------
hostname
influxdb-peers-rc-45i5b
influxdb-peers-rc-dnseu
influxdb-peers-rc-w95vm
influxdb-raft-rc-10o0u
influxdb-raft-rc-m1uyo
influxdb-raft-rc-qi45w
```

PS: Since CLI connects to "http://localhost:8086" by default, I have to use "influx -host <hostIP/name>" when I override the HTTP_BIND_ADDRESS to the host ip.

**Comment by rossmcdonald** [comment]
@supershal Regarding the different ID's, that is to be expected since the meta and data node designations are independent of each other.

Regarding:

> Since CLI connects to "http://localhost:8086" by default, I have to use "influx -host " when I override the HTTP_BIND_ADDRESS to the host ip.

This is also to be expected, since `influxd` is no longer listening on `localhost`. We are hoping to have this configuration issue resolved, but in the meantime you can get around it with tweaks to your `/etc/hosts` file. For example, if you use hostnames instead of IP addresses (`influxN`, where `N` is each node in the cluster) then an `/etc/hosts` file of:

```
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4

0.0.0.0 influx1
```

And set the bind addresses to `influx1:<port>`, then the `influx` command will work as expected.

**Comment by supershal** [comment]
Thanks @rossmcdonald. 
 As shown above, all data nodes (node 4, 5 and 6) also showing in the meta nodes list. Is that expected?
As per clustering doc, raft cluster is form of at least three nodes and rest of nodes are data nodes. I want to keep meta/raft nodes and data nodes on different type of hardwares as per guidelines. What do i need to configure differently when bringing up data nodes such that they do not participate in raft.

**Comment by rossmcdonald** [comment]
@supershal Regarding:

> As shown above, all data nodes (node 4, 5 and 6) also showing in the meta nodes list. Is that expected?

That is correct, since they are both meta nodes and data nodes.

> As per clustering doc, raft cluster is form of at least three nodes and rest of nodes are data nodes. 

This isn't quite right. You need at least three meta nodes to form a healthy raft cluster, but all meta nodes participate in the raft consensus protocol, so they are all technically part of the raft cluster. Let me know if that contradicts the documentation, and I'd be happy to fix it.

> What do i need to configure differently when bringing up data nodes such that they do not participate in raft.

To make a node _only_ a meta node, you have to disable the `[data]` plugin in the [configuration file](https://github.com/influxdata/influxdb/blob/master/etc/config.sample.toml#L49). Similarly, to make a node _only_ a data node, you have to disable the `[meta]` plugin in the [configuration file](https://github.com/influxdata/influxdb/blob/master/etc/config.sample.toml#L25).

**Comment by supershal** [comment]
Thanks @rossmcdonald . This information helped to create separate meta and data cluster as documented. I think this scenario should be mentioned explicitly in the documentation since its important with respect to operational and scaling aspects, specially when the recommendation is to have meta node and data nodes on separate hardwares in the cluster.

also, I noticed that data node's snapshot service is trying to reach meta node at port 8091 at startup. If we disable [meta] plugin in the data nodes then port 8091 is unreachable. So I assume all data nodes should have meta nodes as their peers at startup (-join <metanode1:8091t>, <metanode2>:8091) in order to register with cluster. If thats correct then it would be nice to have it mentioned in documentation.

PS: We are trying to setup influxdb cluster locally and in EC2 using Kubernetes and test our production scenarios. If anyone want to try it out the source code is at: https://github.com/supershal/k8s-influxdb . we will share results once completed.
